src/software/ai/passing/receiver_position_generator.hpp

```diff
@@ -0,0 +1,344 @@
+        {
+            auto pass = Pass::fromDestReceiveSpeed(
+                pass_origin,
+                Point(x_distribution(random_num_gen_), y_distribution(random_num_gen_)),
+                passing_config_);
+            double rating = rateReceivingPosition(world, pass, passing_config_);
+
+            if (rating > best_pass_for_receiving.rating)
+            {
+                best_pass_for_receiving = PassWithRating{pass, rating};
+            }
+        }
+
+        best_receiving_positions.insert_or_assign(zone_id, best_pass_for_receiving);
+    }
+}
+
+template <class ZoneEnum>
+std::vector<ZoneEnum> ReceiverPositionGenerator<ZoneEnum>::getTopZones(
+    const std::map<ZoneEnum, PassWithRating> &best_receiving_positions,
+    unsigned int num_positions, const Point &pass_origin,
+    const std::vector<Point> &existing_receiver_positions)
+{
+    std::vector<ZoneEnum> top_zones;
+
+    // Sort the zones based on initial ratings
+    auto all_zones = pitch_division_->getAllZoneIds();
+    std::sort(all_zones.begin(), all_zones.end(),
+              [&](const ZoneEnum &z1, const ZoneEnum &z2) {
+                  return best_receiving_positions.find(z1)->second.rating >
+                         best_receiving_positions.find(z2)->second.rating;
+              });
+
+    // Iterate through the zones in descending order of rating and select them as top
+    // zones if they are not too close to the previous selected zones.
+    const Angle min_angle_diff_between_receivers =
+        Angle::fromDegrees(passing_config_.receiver_position_generator_config()
+                               .min_angle_between_receivers_deg());
+
+    for (unsigned int i = 0; i < all_zones.size() && top_zones.size() < num_positions;
+         i++)
+    {
+        Angle curr_pass_angle =
+            best_receiving_positions.find(all_zones[i])->second.pass.passerOrientation();
+
+        // Check that none of the previously selected top zones are close to the current
+        // candidate zone
+        bool no_prev_receivers_close =
+            std::none_of(top_zones.begin(), top_zones.end(), [&](const ZoneEnum &zone) {
+                return curr_pass_angle.minDiff(best_receiving_positions.find(zone)
+                                                   ->second.pass.passerOrientation()) <
+                       min_angle_diff_between_receivers;
+            });
+
+        // and none of the existing receiver positions are close to the current
+        // candidate zone
+        no_prev_receivers_close =
+            no_prev_receivers_close &&
+            std::none_of(
+                existing_receiver_positions.begin(), existing_receiver_positions.end(),
+                [&](const Point &existing_receiver_position) {
+                    return curr_pass_angle.minDiff(
+                               (existing_receiver_position - pass_origin).orientation()) <
+                           min_angle_diff_between_receivers;
+                });
```

Comment:
what if you had a receiver on two sides of the field at the exact same (but opposing) y coordinate passing from the origin? `minDiff` wouldn't be able to differentiate that right?

Reply:
You're correct that `minDiff` would return the same angle for both. But I think that is expected since the acute angle between both receivers is the same. Right?

Reply:
So I was thinking that a zone that has the same `minDiff` on the opposite side should be allowed to be added but now that I look at it again, your implementation makes sense